File: 5th-Semester/Design-Pattern/refactoring-workshop/js/src/plaintext_to_html_converter.js
// Adapted from https://github.com/emilybache/Racing-Car-Katas
import fs from 'fs'
import path from 'path'

class PlaintextToHtmlConverter {
    toHtml() {
        const text = fs.readFileSync(path.join(__dirname, 'sample.txt'), 'utf8')
        const htmlLines = this._basicHtmlEncode(text)
        return htmlLines
    }

    _basicHtmlEncode (source) {
        //pick the character from source string
        //and increment the pointer
        var stashNextCharacterAndAdvanceThePointer = () => {
            var c = source.charAt(i)
            i += 1
            return c

        };

        //stringfy convertedLine array and push into result
        //reset convertedLine
        var addANewLine = () => {
            convertedLine = convertedLine.join('')
            result.push(convertedLine)
            convertedLine = []
        };

        var pushACharacterToTheOutput = () => {
            convertedLine.push(characterToConvert);
        };

        var i = 0;
        var result = [];
        var convertedLine = [];
        var characterToConvert = stashNextCharacterAndAdvanceThePointer();
        while (i <= source.length) {

            switch (characterToConvert) {
                case '<':
                    convertedLine.push('&lt;');
                    break;
                case '>':
                    convertedLine.push('&gt;');
                    break;
                case '&':
                    convertedLine.push('&amp;');
                    break;
                case '\n':
                    addANewLine();
                    break;
                default:
                    pushACharacterToTheOutput();
            }

            characterToConvert = stashNextCharacterAndAdvanceThePointer();
        }
        addANewLine();
        result = result.join('<br />');
        return result;
    }
}

export { PlaintextToHtmlConverter }

File: 5th-Semester/Design-Pattern/refactoring-workshop/ruby/lib/plaintext_to_htmlconverter.rb
# Adapted from https://github.com/emilybache/Racing-Car-Katas

class PlaintextToHtmlConverter
  def to_html
    @source = IO.read('./lib/simple.txt')
    basic_html_encode
  end

  private

  def stash_next_character_and_advance_the_pointer
    c = @source[@i]
    @i += 1
    c
  end

  def add_a_new_line
    @converted_line = @converted_line.join('')
    @result.push(@converted_line)
    @converted_line = []
  end

  def push_a_character_to_the_output
    @converted_line.push(@character_to_convert)
  end

  def basic_html_encode
    @i = 0
    @result = []
    @converted_line = []
    @character_to_convert = stash_next_character_and_advance_the_pointer

    while @i <= @source.length
      case (@character_to_convert)
      when "<"
        @converted_line.push("&lt;")
      when ">"
        @converted_line.push("&gt;")
      when "&"
        @converted_line.push("&amp;")
      when "\n"
        add_a_new_line
      else
        push_a_character_to_the_output
      end

      @character_to_convert = stash_next_character_and_advance_the_pointer
    end

    add_a_new_line
    @result = @result.join('<br />')
    @result
  end

end